### DISCUSSION TITLE
Web3Modal V2 - Custom Button + Persist Connection

### DISCUSSION THREAD

**Discussion by saintmaxi** [WalletConnect/walletconnect-monorepo#1556] [Developer Support]
1. for V2, is it possible to use my own button to trigger the modal? if so, what function should i call to trigger the modal?
or is the alternative just to style the default w3m-connect-button?

2. is there a way to allow for a persistent connection across reloads? for example, if my metamask is already connected, to be able to access that without reprompting the modal? similar to this question from v1, 
https://github.com/WalletConnect/web3modal/issues/388

**Comment by xzilja** [comment]
Hey @saintmaxi 

1. Yes, just use open and close functions from [useConnectModal](https://github.com/WalletConnect/web3modal/blob/V2/docs/react.md#useconnectmodal) hook

2. I'm not sure what you mean here, if user connects they will remain connected until they or you trigger disconnect event. You can use [useAccount](https://github.com/WalletConnect/web3modal/blob/V2/docs/react.md#useaccount-example) hook to check for user details and connection status.

If you are not using react, we started writing up some docs about vanilla html/js usage [here](https://github.com/WalletConnect/web3modal/blob/V2/docs/html.md) These are still being worked on, but show all concepts like event subscriptions / controller usage etc.

**Reply by saintmaxi** [reply]
appreciate the reply! the suggestion for #1 worked great, using the vanilla js usage and not sure how I missed the ModalCtrl section haha.

For #2, the behavior i see is that when I am connected and reload the page, it's as if I've lost connection. Though I am still connected as confirmed by Metamask, getting the AccountCtrl state shows address as undefined and connected as false. This is similar to behavior in V1 that I included the issue link for, where you had to essentially trigger the modal automatically to reestablish the connection that already exists, but for V2 this does not seem to work.

**Reply by xzilja** [reply]
I think I know what the issue is. If you look in the examples folder there is a minimal html setup. Account will be initially undefined while everything is being initialised, so you need to add event listeners like these

https://github.com/WalletConnect/web3modal/blob/V2/examples/html/src/events.js

1. Listen for client to be initialised
2. Set up other watchers / subscriptions afterwards, in this case for account
3. Update your ui accordingly

**Reply by xzilja** [reply]
You can also just subscribe to initialised event and then try to get an account. I believe it will be defined, but it is best to use events for this as user accounts can change from "outside" actions i.e. user changes or disconnects account in metamask

**Reply by xzilja** [reply]
With time we'll make docs and examples better to reflect this, I will also try to see if I can abstract some of the complexity (like listening to initialized event for example)

**Reply by saintmaxi** [reply]
gotcha. yeah Account is defined after initialized but state is still as isConnected=false and address=undefined with no further account detected til i open the modal and reprompt mm connection

just printing things out like so
```
ClientCtrl.subscribe(({ initialized }) => {
    if (initialized) {
      const account = AccountCtrl.get();
      console.log(account.address, account.isConnected);
      // Listen for account changes
      AccountCtrl.watch((account) => {
        console.log(account)
        if (account.isConnected) {
          console.log(account.address);
        } else {
          console.log(account);
        }
      });
    }
  });
  ```
  
  Output:
  ```
  >> undefined false
  ```

**Reply by xzilja** [reply]
Hmm, if you remove 
```
const account = AccountCtrl.get();
console.log(account.address, account.isConnected);
```
And only rely on watch, does that one log correct data? If that still doesn't work, could you provide a repo with reproduction, I'll have a look 👍

**Reply by saintmaxi** [reply]
it does not seem to show anything. here's the [repo](https://github.com/saintmaxi/web3modal-example), i just stripped it down to the wallet connection stuff. using Svelte framework integrating the vanilla JS solution, README has instructions to run. the relevant code is here in [Landing.svelte](https://github.com/saintmaxi/web3modal-example/blob/main/src/lib/Landing.svelte)

to reproduce, i would connect with injected provider (I'm using MetaMask), then reload the page (without disconnecting MetaMask). You should see the "Connect Wallet" button again, though if you press to open the Modal and select MetaMask again, it will then detect the existing connection and default to it. I'm basically wondering how to force this connection detection myself instead of making users go thru the modal to reestablish connection.

For example, the [demo site from the V2 README](https://web3modal-dev.pages.dev/) showcases this functionality, where I can refresh and the existing connection is already detected and reestablished.

appreciate you looking into this

**Reply by xzilja** [reply]
@saintmaxi you need to set `autoConnect: true` inside of your config for client to auto reconnect / store account. Just tested this within your example and everything worked.

And this way you can `.get()` as soon as client is initialised so you can reflect connected account right away (with watch there is slight delay)

**Reply by saintmaxi** [reply]
wow can't believe i ignored that parameter. thanks so much for the help, works perfectly now
